### ISSUE TITLE
Remove forced materialization at table sources for queries that do not require them

### ISSUE THREAD

**Issue by vcrfxia** [confluentinc/ksql#6750]
Today, ksqlDB adds a no-op `mapValues()` call to all table sources (with the exception of tables that require repartitions prior to joins -- see https://github.com/confluentinc/ksql/pull/6715) in order to prevent table source topics from being used as changelog topics for state stores (see https://github.com/confluentinc/ksql/issues/5673 for context). However, this means all table queries have these state stores (and associated changelog topics) even if the queries do not need state stores, as is the case with filters and projections. We should not introduce state stores for such queries.

We have at least three ways to do this:
- Update the logic in ksqlDB's PlanInfoExtractor to determine when a query needs a state store, and if not, do not force materialization at the table source. (Here's [example code](https://github.com/vcrfxia/ksql/blob/table-source-mat/ksqldb-execution/src/main/java/io/confluent/ksql/execution/plan/PlanInfoExtractor.java) that does this.)
- Introduce the ability to disable the table-source-topic-as-changelog optimization in Kafka Streams, while still enabling the other optimizations. This is not possible today because the Kafka Streams config for enabling topology optimizations is all or nothing. If ksqlDB can disable the table-source-topic-as-changelog optimization while leaving the other optimizations enabled, then we can remove the dummy `mapValues()` call when building table sources all together, as the underlying problem the `mapValues()` call works around will have been fixed.
- Introduce an option for specifying a "materialization hint" to Kafka Streams, in order to say "if a materialization is needed, materialize at this step; if not, do not materialize." This is in contrast to today where materializations are forced; if a Materialized instance is passed at a given step, then a state store will be created even if it is not needed by the downstream topology. If this option existed, then we could update the dummy `mapValues()` call to specify a materialization hint rather than forcing materialization at that step, thereby avoiding the unused state store for queries that do not require materialization.

The first option is dis-preferred as a longterm solution as it feels redundant for ksqlDB to contain logic for figuring out which topologies require materialization and which don't, as Kafka Streams already does this. As such, we'd like to implement one of the latter two suggested functionalities in Kafka Streams instead, and update ksqlDB accordingly.

Note that one of these three changes is required prior to ksqlDB supporting joins on tables that are aggregated in the same query (i.e., certain types of nested queries / subqueries), as the logic in the PlanInfoExtractor today would incorrectly choose not to materialize at the table source step in such cases, and the table source topic would be used as the changelog for the table source state store in the aggregation, exposing us to the Avro schema name bug linked above.

**Comment by agavra** [comment]
**tl;dr** we're not going to pursue this change and are going to accept the extra state store/changelog topic today.

---

After numerous discussions with @mjsax @guozhangwang @vvcephei @MichaelDrogalis @derekjn @colinhicks  (take a deep breather, that was a lot of names) we've decided not to pursue this. There are a few takeaways:

1. This one was the nail in the coffin - this state store is actually useful for something important: not sending redundant tombstones. There was a confusion historically with HAVING (see #3558) and we suspect there would be something similar here.
2. We had originally thought that this was introduced as part of https://github.com/confluentinc/ksql/issues/5673 but this is not the case. It was there all along, it was just materialized at the source step instead of the mapValues. You can check out the historical plans, which confirm this.
3. The We don't actually know how many users are running into this. They'd have to be creating tables derived from tables, which we suspect isn't something that's super common. If we have evidence that it is, we can reconsider in the future.

That being said, we still think that the third option you suggest "Introduce an option for specifying a "materialization hint" to Kafka Stream" is valuable. Especially because it will allow us to optimize things like "don't materialize the source table, but rather the _result_ of the filter" if one such thing exists. This will be more complicated and would also require streams changes.
